Sources/SwiftParserDiagnostics/LexerDiagnosticMessages.swift

```diff
@@ -325,6 +325,50 @@ extension SwiftSyntax.TokenDiagnostic {
         changes.append(.replaceLeadingTrivia(token: nextToken, newTrivia: []))
       }
       return [FixIt(message: .removeExtraneousWhitespace, changes: changes)]
+    case .spaceAtStartOfRegexLiteral:
+      var ancestor = Syntax(token)
+      while ancestor.kind != .regexLiteralExpr, let parent = ancestor.parent {
+        ancestor = parent
+      }
+
+      let regexLiteral = ancestor.cast(RegexLiteralExprSyntax.self)
+      let regexText = regexLiteral.regex.text
+      let lastIndex = regexText.index(before: regexText.endIndex)
+      let escapedRegexText: String =
+        if regexText.startIndex != lastIndex && regexText[lastIndex].isWhitespace {
```

Comment:
Can we ever hit this? Isn’t the error about the regex *staring* with a space but you’re checking if it ends with a space. Also would be good to check that this Fix-It behaves correctly when having a regex that starts and ends with a space.

---

`isWhitespace` would also be satisfied for characters other than a space in which case we’d be translating the arbitrary whitespace to a space below. Instead of hardcoding the space there, I would just include `regexText[lastIndex]`.

Reply:
There are two possible cases where the regex lexer might emit the error `.spaceAtEndOfRegexLiteral`, however in each case some other error takes precedence over it,

1. `.spaceAtStartOfRegexLiteral` as in `/  /` (both leading and trailing spaces)
```swift
assertParse(
  """
  /1️⃣  /
  """,
  diagnostics: [
    DiagnosticSpec(
      message: "bare slash regex literal may not start with space",
      fixIts: [
        "convert to extended regex literal with '#'",
        #"insert '\'"#,
      ]
    )
  ],
  applyFixIts: ["convert to extended regex literal with '#'"],
  fixedSource: """
    #/  /#
    """
)
```
The current design will just shadow `.spaceAtEndOfRegexLiteral` at the presence of `.spaceAtStartOfRegexLiteral`.
```swift

builder.recordPatternError(.spaceAtStartOfRegexLiteral, at: cursor)
//
// some further regex lexing
//
builder.recordPatternError(.spaceAtEndOfRegexLiteral, at: lastUnespacedSpaceOrTab)
```
because of this
```swift
extension RegexLiteralLexemes.Builder {
 mutating func recordPatternError(
    _ kind: TokenDiagnostic.Kind,
    at cursor: Lexer.Cursor
  ) {
    precondition(lastLexemeKind == .openingSlash)

    // Only record if we don't already have a pattern error, we want to prefer
    // the first error we encounter.
    if patternError == nil {
      patternError = .init(kind, position: cursor)
    }
  }
}
```
2. `missingNode` as in `/a /` (trailing-space only)
```swift
assertParse(
  """
  /a /1️⃣
  """,
  diagnostics: [
    DiagnosticSpec(message: "expected expression after operator", fixIts: ["insert expression"])
  ],
  fixedSource: """
    /a / <#expression#>
    """
)
```

I decided to make the "insert '\'" fix-it also escape the trailing space in case 1 because escaping only the leading space will turn the fixed regex into case 2, e.g. `/  /` -> `/\  /`, that would be undesirable.

In a previous iteration, I've tried not emitting the "insert `\'" fix-it when there're both leading and trailing spaces in the regex, but for some reason I decided to change the design to the current one.

Reply:
Ah, I see. Thanks for the detailed explanation. That makes perfect sense 👍🏽